src/vmm/src/builder.rs

```diff
@@ -763,39 +767,568 @@ pub mod tests {
     use std::fs::File;
 
     use super::*;
+    use arch::DeviceType;
+    use builder::StartMicrovmError::*;
+    use devices::virtio::TYPE_BLOCK;
+    use kernel::cmdline::Cmdline;
     use polly::event_manager::EventManager;
-
-    #[test]
-    fn test_stdin_wrapper() {
-        let wrapper = SerialStdin(io::stdin());
-        assert_eq!(wrapper.as_raw_fd(), io::stdin().as_raw_fd())
+    use seccomp::SECCOMP_LEVEL_ADVANCED;
+    use std::path::PathBuf;
+    use vmm_config::boot_source::DEFAULT_KERNEL_CMDLINE;
+    use vmm_config::drive::BlockDeviceConfig;
+    use vmm_config::net::NetworkInterfaceConfig;
+
+    struct SerialInput(File);
+    impl io::Read for SerialInput {
+        fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+            self.0.read(buf)
+        }
+    }
+    impl AsRawFd for SerialInput {
+        fn as_raw_fd(&self) -> RawFd {
+            self.0.as_raw_fd()
+        }
     }
+    impl devices::legacy::ReadableFd for SerialInput {}
 
-    #[test]
-    fn test_setup_serial_device() {
-        // Wrapper over TempFile that implements `ReadableFd`.
-        struct SerialInput(File);
-        impl io::Read for SerialInput {
-            fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
-                self.0.read(buf)
-            }
+    fn default_kernel_file() -> File {
+        let path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+        let kernel_path = path.parent().unwrap().join("kernel/src/loader");
+
+        #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+        {
+            File::open(kernel_path.join("test_elf.bin")).unwrap()
         }
-        impl AsRawFd for SerialInput {
-            fn as_raw_fd(&self) -> RawFd {
-                self.0.as_raw_fd()
-            }
+
+        #[cfg(target_arch = "aarch64")]
+        {
+            File::open(kernel_path.join("test_pe.bin")).unwrap()
         }
-        impl devices::legacy::ReadableFd for SerialInput {};
+    }
 
+    fn default_mmio_device_manager() -> MMIODeviceManager {
+        MMIODeviceManager::new(
+            &mut (arch::MMIO_MEM_START as u64),
+            (arch::IRQ_BASE, arch::IRQ_MAX),
+        )
+    }
+
+    #[cfg(target_arch = "x86_64")]
+    fn default_portio_device_manager(event_manager: &mut EventManager) -> PortIODeviceManager {
         let read_tempfile = TempFile::new("/tmp/serial_read_").unwrap();
         let read_file = File::open(read_tempfile.as_path()).unwrap();
         let read_handle = SerialInput(read_file);
+
+        let serial =
+            setup_serial_device(event_manager, Box::new(read_handle), Box::new(io::stdout()))
```

Comment:
I would also keep the original test for setting up the serial device.

With this code, it is still tested, but if we were to introduce a bug there, instead of explicitly failing a `test_serial_setup` test we'd only fail tests like `test_attach_net_devices` and `test_attach_block_devices`.